Don't perform any new queries while reading a query result on disk

Aaron1011 opened the pull request:
In addition to being very confusing, this can cause us to add dep node edges between two queries that would not otherwise have an edge.

We now panic if any new dep node edges are created during the deserialization of a query result. This requires serializing the full `AdtDef` to disk, instead of just serializing the `DefId` and invoking the `adt_def` query during deserialization.

I'll probably split this up into several smaller PRs for perf runs.

rust-highfive commented:
r? @jackh726

(rust-highfive has picked a reviewer for you, use r? to override)

Aaron1011 commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

bors commented:
:hourglass: Trying commit 193372013836ac4e5940ebfe088bdd44cdcee7e5 with merge 535f5db4e39c1538a165cd4e32d36e737c439cfd...
<!-- homu: {"type":"TryBuildStarted","head_sha":"193372013836ac4e5940ebfe088bdd44cdcee7e5","merge_sha":"535f5db4e39c1538a165cd4e32d36e737c439cfd"} -->

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/4522833396?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/193372013836ac4e5940ebfe088bdd44cdcee7e5/checks/4522833396/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   |
19 | use rustc_hir::def_id::DefId;
   |     ^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: `-D unused-imports` implied by `-D warnings`

error: unused imports: `Encodable`, `Encoder`
  --> compiler/rustc_middle/src/ty/adt.rs:14:29
   |
14 | use rustc_serialize::{self, Encodable, Encoder};

error: unused import: `ptr`
  --> compiler/rustc_middle/src/ty/adt.rs:23:11
   |
   |
23 | use std::{ptr, str};

error: could not compile `rustc_middle` due to 3 previous errors
warning: build failed, waiting for other jobs to finish...
error: build failed
```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/4523012063?check_suite_focus=true)
<!-- homu: {"type":"TryBuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/4523012063?check_suite_focus=true","builder_name":"checks-actions"} -->

rust-log-analyzer commented:
The job **`dist-x86_64-linux`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/runs/4522836899?check_suite_focus=true) [(plain)](https://github.com/rust-lang-ci/rust/commit/535f5db4e39c1538a165cd4e32d36e737c439cfd/checks/4522836899/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   |
19 | use rustc_hir::def_id::DefId;
   |     ^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: `-D unused-imports` implied by `-D warnings`

error: unused imports: `Encodable`, `Encoder`
  --> compiler/rustc_middle/src/ty/adt.rs:14:29
   |
14 | use rustc_serialize::{self, Encodable, Encoder};

error: unused import: `ptr`
  --> compiler/rustc_middle/src/ty/adt.rs:23:11
   |
   |
23 | use std::{ptr, str};

[RUSTC-TIMING] rustc_query_system test:false 5.896
[RUSTC-TIMING] rustc_session test:false 9.030
[RUSTC-TIMING] rustc_llvm test:false 0.269
```

</details>

Aaron1011 commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

bors commented:
:hourglass: Trying commit b9c62879fd4c4a31eb2fa8606dd0e2351dba51fd with merge 861c72a96000618f65e5a7f8cf224b77be30b4fc...
<!-- homu: {"type":"TryBuildStarted","head_sha":"b9c62879fd4c4a31eb2fa8606dd0e2351dba51fd","merge_sha":"861c72a96000618f65e5a7f8cf224b77be30b4fc"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/4526929959?check_suite_focus=true)
Build commit: 861c72a96000618f65e5a7f8cf224b77be30b4fc (`861c72a96000618f65e5a7f8cf224b77be30b4fc`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/4526929959?check_suite_focus=true"},"merge_sha":"861c72a96000618f65e5a7f8cf224b77be30b4fc"} -->

rust-timer commented:
Queued 861c72a96000618f65e5a7f8cf224b77be30b4fc with parent 404c8471aba60c2d837fa728e7c729a0f52d5830, future [comparison URL](https://perf.rust-lang.org/compare.html?start=404c8471aba60c2d837fa728e7c729a0f52d5830&end=861c72a96000618f65e5a7f8cf224b77be30b4fc).

rust-timer commented:
Finished benchmarking commit (861c72a96000618f65e5a7f8cf224b77be30b4fc): [comparison url](https://perf.rust-lang.org/compare.html?start=404c8471aba60c2d837fa728e7c729a0f52d5830&end=861c72a96000618f65e5a7f8cf224b77be30b4fc).

**Summary**: This change led to very large relevant mixed results 🤷 in compiler performance.
- Very large improvement in instruction counts (up to -24.9% on `incr-unchanged` builds of `issue-46449`)
- Very large regression in instruction counts (up to 16.1% on `incr-full` builds of `webrender-wrench`)

If you disagree with this performance assessment, please file an issue in [rust-lang/rustc-perf](https://github.com/rust-lang/rustc-perf/issues/new).

Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR led to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed.

@bors rollup=never
@rustbot label: +S-waiting-on-review -S-waiting-on-perf +perf-regression

teor2345 commented:
> **Summary**: This change led to very large relevant mixed results 🤷 in compiler performance.
>  * Very large improvement in instruction counts (up to -24.9% on `incr-unchanged` builds of `issue-46449`)
>  * Very large regression in instruction counts (up to 16.1% on `incr-full` builds of `webrender-wrench`)

Seems like an overall performance win, except for a few large outliers.

Aaron1011 commented:
The perf results are very similar to PR https://github.com/rust-lang/rust/pull/91924 which was split out from this one, so I think that PR is responsible for all of the significant performance changes.

lcnr commented on compiler/rustc_middle/src/ty/adt.rs:
```diff
@@ -117,7 +117,7 @@ impl PartialEq for AdtDef {
     // `AdtDef`s are always interned, and this is part of `TyS` equality.
     #[inline]
```
are the `Eq` and `Hash` impls relevant for performance?

because if so, an alternative would be use a different type when dealing with not interned  `AdtDef`s, e.g. by using
```rust
pub struct AdtDef<const INTERNED = true> {
    ...
}
```

lcnr replied:
considering that even all `full` benches are faster now (which i don't really understand), it doesn't seem to matter too much :thinking:

Aaron1011 replied:
I think the change in `full` benchmarks is due to the fact that we encode `AdtDef`s into the crate metadata (as a part of `TyKind`)

lcnr replied:
ah, so it's a tradeoff between the size of the metadata and having to recompute `adt_def` for adts in dependencies when compiling a crate

Aaron1011 replied:
That's right. Surprisingly, this is actually a net disk space win for incr comp in many cases (I checked a few on the benchmarks). The metadata becomes larger, but that's outweighed by the dep graph becoming much smaller.

Aaron1011 replied:
Also, let's move this conversation to https://github.com/rust-lang/rust/pull/91924, which contains just the `AdtDef` changes

jackh726 commented:
Marking as blocked on #91924

michaelwoerister reviewed (commented):
I added some comments below.

Aaron1011 commented on compiler/rustc_query_system/src/query/plumbing.rs:
```diff
@@ -515,7 +515,13 @@ where
     if query.cache_on_disk {
         let prof_timer = tcx.dep_context().profiler().incr_cache_loading();
-        let result = query.try_load_from_disk(tcx, prev_dep_node_index);
+
```
We're still going to go with the approach of banning all new `DepNode`s, right? In that case, I think I can simplify the comment to just mention the reasoning for not creating any new `DepNode`s.

Aaron1011 commented:
@michaelwoerister: I've moved the construction of the new `TaskDeps`, and added a detailed comment. I didn't add a new check for the node being green, since we already check that in `try_load_from_disk_and_cache_in_memory` (regardless of whether we actually deserialize from disk).

michaelwoerister commented on compiler/rustc_query_system/src/dep_graph/graph.rs:
```diff
@@ -171,6 +171,57 @@ impl<K: DepKind> DepGraph<K> {
+    /// `A -> B -> C`
+    ///
+    /// Suppose that decoding the result of query `B` required invoking
+    /// a query `D`. If we did not create a fresh `TaskDeps` when
```
I would prefer if we changed that to

> Suppose that decoding the result of query `B` required re-computing query `C`.

That would be OK in principle since it would be compatible with the dep-graph we already created. But creating any edge away from `B` will never be legal (not even if we allowed query invocation during cache loading) because we already allocated `B`'s dep-node during red-green marking. The current wording makes it sound like it would be OK for a "new" query to be called during decoding.

Aaron1011 replied:
My thinking was to explain the process for an arbitrary new node, and then explain that this can only work if the node already exists. However, I'm fine with changing the wording.

michaelwoerister commented on compiler/rustc_query_system/src/dep_graph/graph.rs:
```diff
@@ -171,6 +171,57 @@ impl<K: DepKind> DepGraph<K> {
+    ///
+    /// Suppose that decoding the result of query `B` required invoking
+    /// a query `D`. If we did not create a fresh `TaskDeps` when
+    /// decoding `B`, we might would still be using the `TaskDeps` for query `A`
```
~~might~~

michaelwoerister commented on compiler/rustc_query_system/src/dep_graph/graph.rs:
```diff
@@ -171,6 +171,57 @@ impl<K: DepKind> DepGraph<K> {
+    /// a query `D`. If we did not create a fresh `TaskDeps` when
+    /// decoding `B`, we might would still be using the `TaskDeps` for query `A`
+    /// (if we needed to re-execute `A`). This would cause us to create
+    /// a new edge `A -> D`. If this edge did not previously
```
this would then become `A -> C`

michaelwoerister commented on compiler/rustc_query_system/src/dep_graph/graph.rs:
```diff
@@ -171,6 +171,57 @@ impl<K: DepKind> DepGraph<K> {
+    /// get an ICE. Normally, we would have tried (and failed) to mark
+    /// some other query green (e.g. `item_children`) which was used
+    /// to obtain `D`, which would prevent us from ever trying to force
+    /// a non-existent `D`.
```
It should be fine to keep the rest of the paragraph as is, just with `D` replaced by `C`.

michaelwoerister commented on compiler/rustc_query_system/src/dep_graph/graph.rs:
```diff
@@ -171,6 +171,57 @@ impl<K: DepKind> DepGraph<K> {
+    /// of `B`, this would result in an edge `B -> D`. If that edge already
+    /// existed (with the same `DepPathHash`es), then it should be correct
+    /// to allow the invocation of the query to proceed during deserialization
+    /// of a query result. However, this would require additional complexity
```
Maybe add a sentence before "However, ...":

> We would merely assert that the dep-graph fragment that would have been added by invoking `C` while decoding `B` conforms to the dep-graph fragment we already instantiated for `B` at the point where we successfully marked `B` as green.

Aaron1011 replied:
I added this comment with the wording slightly tweaked ("equivalent to" instead of "conforms to")

michaelwoerister reviewed (commented):
I added some suggestions for improving the comment. I know I'm being a bit of a stickler here -- but this is complicated stuff and I want things to be as clear as possible, especially for people who come across this but are not experts on the query system.

Aaron1011 commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

bors commented:
:hourglass: Trying commit bfe15608a12813a3743e3cacc5ed529fbeed63a5 with merge 8ed49fb7ec8f30d7dd69865ae1382fbc31ceb412...
<!-- homu: {"type":"TryBuildStarted","head_sha":"bfe15608a12813a3743e3cacc5ed529fbeed63a5","merge_sha":"8ed49fb7ec8f30d7dd69865ae1382fbc31ceb412"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/4601667826?check_suite_focus=true)
Build commit: 8ed49fb7ec8f30d7dd69865ae1382fbc31ceb412 (`8ed49fb7ec8f30d7dd69865ae1382fbc31ceb412`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/4601667826?check_suite_focus=true"},"merge_sha":"8ed49fb7ec8f30d7dd69865ae1382fbc31ceb412"} -->

rust-timer commented:
Queued 8ed49fb7ec8f30d7dd69865ae1382fbc31ceb412 with parent e100ec5bc7cd768ec17d75448b29c9ab4a39272b, future [comparison URL](https://perf.rust-lang.org/compare.html?start=e100ec5bc7cd768ec17d75448b29c9ab4a39272b&end=8ed49fb7ec8f30d7dd69865ae1382fbc31ceb412).

rust-timer commented:
Finished benchmarking commit (8ed49fb7ec8f30d7dd69865ae1382fbc31ceb412): [comparison url](https://perf.rust-lang.org/compare.html?start=e100ec5bc7cd768ec17d75448b29c9ab4a39272b&end=8ed49fb7ec8f30d7dd69865ae1382fbc31ceb412).

**Summary**: This change led to large relevant regressions 😿 in compiler performance.
- Large regression in instruction counts (up to 1.1% on `incr-full` builds of `externs`)

If you disagree with this performance assessment, please file an issue in [rust-lang/rustc-perf](https://github.com/rust-lang/rustc-perf/issues/new).

Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR led to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed.

@bors rollup=never
@rustbot label: +S-waiting-on-review -S-waiting-on-perf +perf-regression

michaelwoerister commented:
I'm not sure I trust those performance numbers. Most of the listed regressions are `incr-full` cases -- which should not hit the code path that is modified by this PR.

Anyway, even if the numbers are real, I don't see how this correctness fix could be implemented in a significantly cheaper way.

pnkfelix commented:
discussed in T-compiler meeting. @Aaron1011 says that this is coupled to other PR's; it cannot be backported on its own to beta.

Deferring decision about beta-backport until @michaelwoerister is able to address the concerns raised by @Aaron1011 (or, in absence of such communications, we will default to not beta-backporting, which should be relatively safe we think).

michaelwoerister commented:
As mentioned in the triage meeting, backporting this PR would indeed require https://github.com/rust-lang/rust/pull/91924 to be backported too. Since that PR seems to be causing https://github.com/rust-lang/rust/issues/92163 and possibly other issues, backporting either does not seem to be an option.

Aaron1011 commented:
@michaelwoerister: I've adjusted the comment as you requested

BGR360 commented:
@rustbot label -S-waiting-on-author +S-waiting-on-review

apiraino commented:
Beta backport declined as per compiler team discussion [on Zulip](https://rust-lang.zulipchat.com/#narrow/stream/238009-t-compiler.2Fmeetings/topic/.5Bweekly.5D.202021-12-30.20.2354818/near/266435888)

@rustbot label -beta-nominated

cjgillot commented on compiler/rustc_query_system/src/dep_graph/graph.rs:
```diff
@@ -171,6 +171,62 @@ impl<K: DepKind> DepGraph<K> {
+        let mut deps = TaskDeps::default();
+        deps.read_allowed = false;
+        let deps = Lock::new(deps);
+        K::with_deps(Some(&deps), op)
```
Should we prefer to use an enum for deps:
```rust
enum TaskDepsRef<'a> {
  Allow(&'a Lock<TaskDeps>),
  Ignore,
  Forbid,
}
```

Aaron1011 replied:
Opened https://github.com/rust-lang/rust/pull/92681

cjgillot commented on compiler/rustc_query_system/src/query/plumbing.rs:
```diff
@@ -515,7 +514,13 @@ where
+        // are created during deserialization. See the docs of that method for more
+        // details.
+        let result = dep_graph
+            .with_query_deserialization(|| query.try_load_from_disk(tcx, prev_dep_node_index));
```
I think we can even forbid queries inside the invocation code. Can this call be moved around the `NotYetStarted` branch of `try_execute_query`? It's not supposed to invoke any user-provided code directly, only call into the `DepGraph` to setup a `TaskDeps`.

Aaron1011 replied:
I tried this earlier, but there are some invocations of `def_span` within the query infrastructure itself. I think those could be removed in a follow-up PR, since this PR should catch almost all of the cases.

cjgillot commented on compiler/rustc_query_system/src/dep_graph/graph.rs:
```diff
@@ -171,6 +171,62 @@ impl<K: DepKind> DepGraph<K> {
+    pub fn with_query_deserialization<OP, R>(&self, op: OP) -> R
+    where
+        OP: FnOnce() -> R,
+    {
```
Should we add a fast path if the dep-graph is disabled?

michaelwoerister replied:
This should never be called, if the dep-graph is disabled, right? Maybe we could add a debug assertion that the dep-graph is indeed enabled.

michaelwoerister commented:
This looks good to me now. @cjgillot's comments make sense to me, but I'd be fine with addressing them at a later point in time.

jackh726 commented:
r? @michaelwoerister

michaelwoerister commented:
Let's get this merged. 

@bors r+

bors commented:
:pushpin: Commit 27ed52c0a2e05d459a25d077d2eec584cb98a591 has been approved by `michaelwoerister`

<!-- @bors r=michaelwoerister 27ed52c0a2e05d459a25d077d2eec584cb98a591 -->
<!-- homu: {"type":"Approved","sha":"27ed52c0a2e05d459a25d077d2eec584cb98a591","approver":"michaelwoerister"} -->

bors commented:
:hourglass: Testing commit 27ed52c0a2e05d459a25d077d2eec584cb98a591 with merge a7e2e33960e95d2eb1a2a2aeec169dba5f73de05...
<!-- homu: {"type":"BuildStarted","head_sha":"27ed52c0a2e05d459a25d077d2eec584cb98a591","merge_sha":"a7e2e33960e95d2eb1a2a2aeec169dba5f73de05"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/4750041040?check_suite_focus=true)
Approved by: michaelwoerister
Pushing a7e2e33960e95d2eb1a2a2aeec169dba5f73de05 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"michaelwoerister","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/4750041040?check_suite_focus=true"},"merge_sha":"a7e2e33960e95d2eb1a2a2aeec169dba5f73de05"} -->

rust-timer commented:
Finished benchmarking commit (a7e2e33960e95d2eb1a2a2aeec169dba5f73de05): [comparison url](https://perf.rust-lang.org/compare.html?start=488acf86a75c56d30b16822e953c505a9e4901a7&end=a7e2e33960e95d2eb1a2a2aeec169dba5f73de05).

**Summary**: This benchmark run did not return any relevant changes.

If you disagree with this performance assessment, please file an issue in [rust-lang/rustc-perf](https://github.com/rust-lang/rustc-perf/issues/new).



@rustbot label: -perf-regression
